index.bs

```diff
@@ -181,12 +182,18 @@ The <a href="https://www.w3.org/TR/webxr/">WebXR Device API API</a> defines each
 
 Augmented Reality devices that use transparent optical elements frequently exhibit {{XREnvironmentBlendMode/"additive"}} blending behavior even for {{XRSessionMode/"immersive-vr"}} sessions.
 
-The {{XRSession/environmentBlendMode}} MUST NOT be {{opaque}} for {{immersive-ar}} sessions. When an [=xr device=] [=performs automatic composition=] using a passthrough camera in an {{XRSessionMode/"immersive-ar"}} session, it MUST report {{XREnvironmentBlendMode/"alpha-blend"}} blending behavior.
+The {{XRSession/environmentBlendMode}} MUST NOT be {{opaque}} for {{immersive-ar}} sessions. When an [=xr device=] [=perform automatic composition|performs automatic composition=] using a passthrough camera in an {{XRSessionMode/"immersive-ar"}} session, it MUST report {{XREnvironmentBlendMode/"alpha-blend"}} blending behavior.
 
 Automatic composition {#automatic-composition}
 ---------
-<dfn>performs automatic composition</dfn>
-TODO add explanation here.
+
+An [=xr device=] will have its [=xr compositor=] <dfn>perform automatic composition</dfn> by drawing the rendered scene on top of the real-world background. Such a process MUST NOT grant the page access to any information about the real world, like camera data or geometry.
+
+In case the [=xr device=] is using a passthrough camera, it MUST use alpha-blending to deal with any transparent pixels in the rendered scene, and MUST have an {{XRSession/environmentBlendMode}} of {{alpha-blend}}. If it is an additive light device where the real-world background is visible through the device's transparent display, it MUST ignore alpha values when performing automatic composition, instead, it MUST treat black pixels as transparent.
```

Comment:
> ignore alpha values

@thetuvix Is this accurate?... it doesn't seem quite right to me, but it's been a while...

Reply:
this is what the existing environmentblendmode says

Reply:
Doesn't mean it's accurate ;) I'm concerned this language implies that an alpha value of 50% will cause the pixel to actually draw and an alpha value of 100%.  If that's the case, we should put stronger language in the explanation.  If not, we probably need to adjust the language a bit.  @thetuvix @rcabanier ?

Reply:
There is no "automatic composition" in an additive display so I think this paragraph needs to be reworded so it only applies to source-over devices

Reply:
This was the discussion we were having earlier, we're considering this to be automatic composition as well, it's just _super_ automatic in that it's backed by the laws of physics, not software.

Reply:
If the UA doesn't do anything, there shouldn't be language in the spec that implies that it does.
For additive, Alpha is just used during the composition of the scene. The original scene is filled with 100% black so the end result should have no alpha values. (Everything is 100% opaque)
The values are then "added" to cololrs of the current world.
black + any color = any color
white + any color = white

Reply:
> The original scene is filled with 100% black so the end result should have no alpha values.

That's an implementation detail, the scene as rendered to the webgllayers may still have alpha values.

Reply:
> > The original scene is filled with 100% black so the end result should have no alpha values.
> 
> That's an implementation detail, the scene as rendered to the webgllayers may still have alpha values.

The end result from the compositor is 100% black. There are no more alpha values when the "automatic" composition with the real world happens.

In reality, because you always draw with premultiplied alpha, you just ignore the alpha channel and send the rgb data to the projector.

Reply:
> There are no more alpha values when the "automatic" composition with the real world happens.

Yes, I don't see how this contradicts the spec.

Reply:
> Yes, I don't see how this contradicts the spec.

The spec says:

> it MUST ignore alpha values when performing automatic composition, instead, it MUST treat black pixels as transparent.

This is not correct.
Before the "automatic" composition in the waveguide happens, the XRCompositor will set all alpha valuesr to 1. 
The "automatic" composition happens next with a compositing mode of `lighter` 
 See https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_plus

So maybe all you need is:
`The automatic composition MUST happen with a compositing mode of lighter `

We should also state that the real world is opaque. It's needed for both blend modes.

Reply:
> Before the "automatic" composition in the waveguide happens, the XRCompositor will set all alpha valuesr to 1.

That's an implementation detail: the alpha values in the _rendered scene_ are ignored.

Reply:
> That's an implementation detail: the alpha values in the _rendered scene_ are ignored.

OK. Let's ignore that for now :-)

Spec should say:

- real world pixels are always opaque (alpha = 100%)
- in devices with a source-over blend mode the xrcompositor will use the source-over compositing mode to blend with real world pixels
- in devices with a additive blend mode the xrcompositor  will use the lighter compositing mode to blend with real world pixels

With this, there's no need to say anything about alpha values of the scene (since the compositing formulas define what happens)

Reply:
We could make this clearer, I'm not sure how to do that well.

Reply:
The lighter compositing mode still includes the alpha of the rendered scene, though?

Reply:
Aside: we're not certain if we want to specify that alpha values are _ignored_, there's some discussion above about some displays potentially supporting including alpha values during the composition. May also call for a new environment blend mode variant.

Reply:
> The lighter compositing mode still includes the alpha of the rendered scene, though?

in theory yes.
However, since we defined that the destination buffer (aka the real world) is always opaque, the alpha values of the scene don't matter because alpha can't go higher than 100%.

A UA can then optimize its render path by ignoring the alpha of the scene. We don't have to call that out in the spec.

Reply:
> May also call for a new environment blend mode variant.

yes, that would be a new blend mode. Those types of devices are still experimental afaik

Reply:
> However, since we defined that the destination buffer (aka the real world) is always opaque, the alpha values of the scene don't matter because alpha can't go higher than 100%.

No, I mean that the lighter formula includes alpha for both scene and background in the color computation:

```
Fa = 1; Fb = 1
co = αs x Cs + αb x Cb;
αo = αs + αb
```

whereas what we want is something that's more like the color "replacing" the other if it is non-black, without any change due to alpha

Reply:
> No, I mean that the lighter formula includes alpha for both scene and background in the color computation:
> 
> ```
> Fa = 1; Fb = 1
> co = αs x Cs + αb x Cb;
> αo = αs + αb
> ```
> whereas what we want is something that's more like the color "replacing" the other if it is non-black, without any change due to alpha

No, there is no "replacing" of color.
Just follow the formula. It defines the correct behavior.
For instance, if you draw 50% red with an alpha of 40% and you look at a surface that is 30% red (rgba = 0.5,0,0,100%) you get:

> co = .4 x .5 + 1 x 0.5 = .7
> αo = .4 + 1 = 1

Result = 70% red

How exactly this is implemented is up to the UA

Reply:
>  For instance, if you draw 50% red with an alpha of 40% and you look at a surface that is 50% red (rgba = 0.5,0,0,100%) you get:

This doesn't match how the additive blend mode is already defined.

Reply:
> This doesn't match how the additive blend mode is already defined.

I think we're talking past each other :-)
Maybe we can have a quicky call?

Reply:
Sure

Reply:
Sent an invite

Reply:
https://github.com/immersive-web/webxr-ar-module/issues/14